### PR TITLE
PA-24295 Inssqs Implementation

### PR DESCRIPTION
**Pull request by rafet**
# inssqs - AWS SQS Package

The `inssqs` package provides an interface and implementation for interacting with Amazon Simple Queue Service (SQS) in Go. It simplifies sending and deleting messages in batches while handling retries, respecting batch size constraints, and supporting concurrent operations.

## Features

- **Batch Operations**: Send and delete messages in batches to SQS queues.
- **Retries**: Automatic retries for failed operations based on a configurable retry count.
- **Concurrency**: Concurrent processing of multiple batches with specified worker count.
- **Error Handling**: Detailed error logging and handling for failed operations.

## Installation

To use this package in your Go project, import it as follows:

```go
import "github.com/useinsider/go-pkg/inssqs"
```

### Initialization
Initialize an SQS queue instance by providing configuration settings using `NewSQS()`
```go
config := inssqs.Config{
    Region:            "your-aws-region",
    QueueName:         "your-queue-name",
    RetryCount:        3,
    MaxBatchSize:      10,
    MaxBatchSizeBytes: 1024,
    MaxWorkers:        5,
    LogLevel:          "info",
}

sqs := inssqs.NewSQS(config)
```
### Sending Messages
Send a batch of messages to an SQS queue using `SendMessageBatch()`

```go
messages := []inssqs.SQSMessageEntry{
    // Create SQS message entries here
}

failedMessages, err := sqs.SendMessageBatch(messages)
if err != nil {
    // Handle error
}
```

### Deleting Messages
Delete a batch of messages from an SQS queue using `DeleteMessageBatch()`

```go
messages := []inssqs.SQSMessageEntry{
    // Create SQS message entries here
}

failedMessages, err := sqs.SendMessageBatch(messages)
if err != nil {
    // Handle error
}
```

## Configuration Options
- Region: AWS region where the SQS queue resides.
- QueueName: Name of the SQS queue.
- RetryCount: Number of retry attempts allowed for queue operations.
- MaxBatchSize: Maximum size of a message batch.
- MaxBatchSizeBytes: Maximum size of a message batch in bytes.
- MaxWorkers: Maximum number of workers for concurrent operations.
- LogLevel: Log level for SQS operations.
For more details on each function and its parameters, refer to the code documentation.

## Contribution
Feel free to contribute by forking this repository and creating pull requests. Please ensure to adhere to the existing code style and conventions.

### CODE REVIEW

File: inssqs/inssqs.go [hunk 1]
```diff
@@ -0,0 +1,402 @@
...
+	// set endpoint url if provided
+	if config.EndpointUrl != "" {
+		cfg.BaseEndpoint = aws.String(config.EndpointUrl)
+	}
+
+	var logger *inslogger.AppLogger
+	if config.LogLevel == "" {
+		logger = inslogger.NewNopLogger()
+	} else {
+		logger = inslogger.NewLogger(inslogger.LogLevel(config.LogLevel))
```
Comment (Jamil-Najafov) [inline inssqs/inssqs.go:79]: I believe not depending on a specific logger implementation would be beneficial. Just printing out request ids and simpler unit testing comes to mind.

File: inssqs/inssqs.go [hunk 2]
```diff
@@ -0,0 +1,402 @@
...
+		wg.Add(1)
+		go func(b []T) {
+			defer wg.Done()
+			concurrentLimiter <- struct{}{}
+			defer func() { <-concurrentLimiter }()
+			fe, err := f(b, retryCount+1)
+			if err != nil {
+				mu.Lock()
+				outerErr = err
+				mu.Unlock()
```
Comment (Jamil-Najafov) [inline inssqs/inssqs.go]: What is the purpose of the mutex lock here?

File: inssqs/helper.go [hunk 3]
```diff
@@ -0,0 +1,37 @@
+package inssqs
+
+import (
+	"encoding/json"
+)
+
+func createBatches[T any](v []T, recordLimit int, byteLimit int) ([][]T, error) {
```
Comment (Jamil-Najafov) [inline inssqs/helper.go]: It'be great reusable function in a seprate package such as collectionhelper, inskinesis also used something similar.

Comment (rafet) [reply inssqs/helper.go]: `insdash` ???

File: inssqs/inssqs.go (hunk 2 above)
Comment (rafet) [reply inssqs/inssqs.go]: tekrar bi bakinca gerek olmadigini farkettim simdi. mutex'lik bi durum yokmus burda. kaldiriyorum

File: inssqs/inssqs.go [hunk 4]
```diff
@@ -0,0 +1,402 @@
...
+
+// SendMessageBatch sends a batch of messages to an SQS queue, handling retries and respecting batch size constraints.
+//
+// Parameters:
+// - entries: A slice of SQSMessageEntry representing messages to be sent in batches to the SQS queue.
+//
+// Returns:
+// - failedEntries: A slice of SQSMessageEntry containing the messages that failed to be sent after all attempts.
+// - err: An error indicating any failure during the sending process, nil if all messages were sent successfully.
+func (q *queue) SendMessageBatch(entries []SQSMessageEntry) ([]SQSMessageEntry, error) {
```
Comment (sezaakgun) [inline inssqs/inssqs.go:133]: `SendMessageBatch` and `DeleteMessageBatch` is working with it's own concurrency limit. Meaning if `MaxWorkers=3` and `SendMessageBatch` is called 3 times concurrently we have 9 workers at the same time. This is okay when we think about our use case but this should mentioned in the docs as a warning note. I also think we should have the non concurrent versions of these functions.

File: inslogger/inslogger.go [hunk 5]
```diff
@@ -0,0 +1,148 @@
+package inslogger
```
Comment (sezaakgun) [inline inslogger/inslogger.go:1]: READE.md pls

File: inssqs/inssqs.go (hunk 1 above)
Comment (rafet) [reply inssqs/inssqs.go:79]: ne oneriyosun burada tam olarak anlamadim. logger'i iptal edip std out yapmak mi

Comment (Jamil-Najafov) [reply inssqs/inssqs.go:79]: Depending on interfaces rather than concrete implementations would be a good  idea.

Comment (rafet) [reply inssqs/inssqs.go:79]: hmm makes sense

File: inssqs/helper.go (hunk 3 above)
Comment (Jamil-Najafov) [reply inssqs/helper.go]: `insdash` would be great if there's no fitting slice library already available. For instance, we're also frequently needing `slice.Contains` which is also lacking in Go stdlib, which we'd expect such a library to provide.

File: inslogger/inslogger.go (hunk 5 above)
Comment (rafet) [reply inslogger/inslogger.go:1]: done

File: inssqs/inssqs.go (hunk 4 above)
Comment (rafet) [reply inssqs/inssqs.go:133]: warn note is done.

actually user can call these functions by non-concurrently by not passing workers parameter. if they don't, workers will be 1 and message batches will be sent sequantial

File: inssqs/inssqs.go (hunk 1 above)
Comment (rafet) [reply inssqs/inssqs.go:79]: done

File: inssqs/helper.go (hunk 3 above)
Comment (rafet) [reply inssqs/helper.go]: _there will be always not-fitting parts to us on 3rd party libraries..._

I created the `insdash`, we can add our helper functions there, incrementally, when we need.

File: inssqs/inssqs.go (hunk 4 above)
Comment (sezaakgun) [reply inssqs/inssqs.go:133]: Yes but what if they want to send one message type concurrently and another one sequentially. I'm aware that this is not our usecase but can be helpful for others maybe. We can leave this as TODO as well btw.

Comment (rafet) [reply inssqs/inssqs.go:133]: ok I understood, I'll note it as TODO

Comment (sezaakgun) [reply inssqs/inssqs.go:133]: Can you also add this note to the readme as well?

**Review by Jamil-Najafov — DISMISSED** [review]

**Review by sezaakgun — DISMISSED** [review]

**Review by sezaakgun — DISMISSED** [review]

File: inssentry/sentry.go [hunk 6]
```diff
@@ -45,6 +45,21 @@ func Error(err error) {
 	sentry.CaptureException(err)
 }
 
+// ErrorWithAdditionalData sends the error to sentry with additional data.
+func ErrorWithAdditionalData(err error, key string, value interface{}) {
+	if !cachedSettings.IsProduction {
+		fmt.Println(err)
```
Comment (Jamil-Najafov) [inline inssentry/sentry.go]: bunun yerine built-in log.Println kullanalım. fmt.Println sadece debuglar için.

Comment (rafet) [reply inssentry/sentry.go]: done

**Review by sezaakgun — DISMISSED** [review]

**Review by Jamil-Najafov — DISMISSED** [review]

**Review by sezaakgun — APPROVED** [review]

**Review by Jamil-Najafov — APPROVED** [review]

**Pull request merged**